stl/inc/shared_mutex

```diff
@@ -313,43 +437,68 @@ public:
         return _Owns;
     }
 
+    _Pre_satisfies_(_Owns == true)
+    _Pre_satisfies_(_Pmtx != nullptr)
+    _Post_satisfies_(_Owns == false)
+    _When_(_Owns == false, _Analysis_noreturn_)
+    _When_(_Pmtx == nullptr, _Analysis_noreturn_)
+    _When_(_Owns == true, _Releases_shared_lock_(*this->_Pmtx))
+    _When_(_Owns == true, _Releases_nonreentrant_lock_(*this->_Pmtx))
     void unlock() { // try to unlock the mutex
         if (!_Pmtx || !_Owns) {
             _Throw_system_error(errc::operation_not_permitted);
         }
 
+        _Analysis_assume_lock_held_(_Pmtx);
         _Pmtx->unlock_shared();
         _Owns = false;
     }
 
+    _Swaps_locks_(*this->_Pmtx, *_Right._Pmtx)
+    _Post_satisfies_(_Owns == _Right._Owns)
+    _Post_satisfies_(_Right._Owns == _Owns)
     void swap(shared_lock& _Right) noexcept {
         _STD swap(_Pmtx, _Right._Pmtx);
         _STD swap(_Owns, _Right._Owns);
     }
 
+    _Detaches_lock_(return, *this->_Pmtx)
+    _Post_satisfies_(_Pmtx == nullptr)
+    _Post_satisfies_(_Owns == false)
     mutex_type* release() noexcept {
         _Mutex* _Res = _Pmtx;
         _Pmtx        = nullptr;
         _Owns        = false;
         return _Res;
     }
 
-    _NODISCARD bool owns_lock() const noexcept {
+    _NODISCARD
+    _Post_satisfies_(return == _Owns)
+    bool owns_lock() const noexcept {
         return _Owns;
     }
 
+    _Post_satisfies_(return == _Owns)
     explicit operator bool() const noexcept {
         return _Owns;
     }
 
-    _NODISCARD mutex_type* mutex() const noexcept {
+    _NODISCARD
+    _Post_satisfies_(return == _Pmtx)
+    mutex_type* mutex() const noexcept {
         return _Pmtx;
     }
 
 private:
     _Mutex* _Pmtx = nullptr;
     bool _Owns    = false;
 
+    _Pre_satisfies_(_Owns == false)
+    _Pre_satisfies_(_Pmtx != nullptr)
+    _Post_satisfies_(_Owns == false)
+    _Post_satisfies_(_Pmtx != nullptr)
+    _When_(_Owns == false, _Analysis_noreturn_)

```

Comment:
Also, I think this line is damaged. `_Owns == false` is both a precondition and a postcondition, yet this is saying that it's `_Analysis_noreturn_`. I think that `_When_(_Owns == true,` was intended here.

Reply:
I think the compiler didn’t warn without mentioning it this way.
Will double check.